### ISSUE TITLE
MSI Z690/Z790: Ability to disable P cores or E cores and Hyperthreading within Dasharo

### ISSUE THREAD

**Issue by xmhd** [Dasharo/dasharo-issues#973]
### The problem you're addressing (if any)

Currently if I use MSI Z690/Z90 with Intel 12700k or such, it has a mixture of P (performance) and E (efficiency) cores. By default the system boots with both enabled. I am currently not able to disable these within Dasharo. This also applies to hyperthreading.

### Describe the solution you'd like

Some BIOS from other vendors give the ability to disable / custom select the number of P cores or E cores in use. It would be nice to have something similar within Dasharo. This also applies to hyperthreading.

### Where is the value to a user, and who might that user be?

User can hard disable efficiency cores and/or hyperthreading within Dasharo and not have to tinker with their operating system to do so.

This would also enable better compatibility with more Unix variants e.g. some BSDs and illumos who do not yet have proper support for Intel Thread Director.

There is also an added security benefit in that hyperthreading can be disabled in an out-of-the-box solution for security focused releases of Dasharo.

### Describe alternatives you've considered

Disabling within the operating system. Unfortunately depends on the operating system supporting such a thing.

**Comment by macpijan** [comment]
I assume you meant for MSI boards? It actually has been implemented already on another platform, and we plan to have it integrated to the future MSI release as well.

**Comment by pietrushnic** [comment]
Hi @xmhd, thanks for the report.

> Some BIOS from other vendors give the ability to disable / custom select the number of P cores or E cores in use.

My $0.02:
I guess none of those BIOSes are open-source. There are many requests to add more features to Dasharo. Still, a comparison to closed source BIOSes developed over the last 30 years with a business model that gives a project budget at the level of 100x+ in contrast to what we have done is slightly unfair. A closed-source BIOS feature set cannot drive us. We don't compete in the same market. Our market is for those who value openness and trustworthiness. If one values a lot of features, one can choose a source option or support open-source with the wallet, so open-source can make closed-source accountable.

**Comment by xmhd** [comment]
> I assume you meant for MSI boards? It actually has been implemented already on another platform, and we plan to have it integrated to the future MSI release as well.

Yes indeed, I will update the main post to reflect that.

> Hi @xmhd, thanks for the report.
> 
> > Some BIOS from other vendors give the ability to disable / custom select the number of P cores or E cores in use.
> 
> My $0.02: I guess none of those BIOSes are open-source. There are many requests to add more features to Dasharo. Still, a comparison to closed source BIOSes developed over the last 30 years with a business model that gives a project budget at the level of 100x+ in contrast to what we have done is slightly unfair. A closed-source BIOS feature set cannot drive us. We don't compete in the same market. Our market is for those who value openness and trustworthiness. If one values a lot of features, one can choose a source option or support open-source with the wallet, so open-source can make closed-source accountable.

I'm aware of the fact that none of the other BIOSes are open-source, and their proprietary nature as a whole within the industry. I am merely providing a comparison. I also find it odd to say that a closed-source BIOS feature set cannot drive us, yet that is exactly what Dasharo currently provides e.g. XMP memory overclocking, and you yourself had a different opinion to that feature request.

**Comment by xmhd** [comment]
> I assume you meant for MSI boards? It actually has been implemented already on another platform, and we plan to have it integrated to the future MSI release as well.

I would be happy to help beta test such a feature by the way :-)

**Comment by pietrushnic** [comment]
AFAIK XMP happens to be a disaster on multiple fronts. I'm just saying that we cannot and right now don't want to compete in a way that "we have more features than BIOS Vendor X." It is clueless. FSP has thousands of options that could be exposed. More options mean more problems in support, and there is a limit to what the Dasharo Team can provide, and we still have time to develop something. According to @macpijan this feature request is valid and was already implemented.

Some time ago, there was a discussion about that on Matrix; my position is that Dasharo needs one thing that would do better than any other vendor (maybe UEFI Secure Boot), and we should focus more on that than having a gazillion features which only cause headaches to support channels. Moreover, this is open-source, and only skills limit the community from providing whatever option they want. By looking at the BIOS modding community, people go into much, much more hassle than coding to enable some exotic features. 

I'm not saying this applies here. My rant is only because of the comparison to the proprietary vendors, which is a comparison at the scale of 1 to 400 in terms of developers and 1 to millions in comparison to market cap.

**Comment by xmhd** [comment]
> AFAIK XMP happens to be a disaster on multiple fronts. I'm just saying that we cannot and right now don't want to compete in a way that "we have more features than BIOS Vendor X." It is clueless. FSP has thousands of options that could be exposed. More options mean more problems in support, and there is a limit to what the Dasharo Team can provide, and we still have time to develop something. According to @macpijan this feature request is valid and was already implemented.
> 
> Some time ago, there was a discussion about that on Matrix; my position is that Dasharo needs one thing that would do better than any other vendor (maybe UEFI Secure Boot), and we should focus more on that than having a gazillion features which only cause headaches to support channels. Moreover, this is open-source, and only skills limit the community from providing whatever option they want. By looking at the BIOS modding community, people go into much, much more hassle than coding to enable some exotic features.
> 
> I'm not saying this applies here. My rant is only because of the comparison to the proprietary vendors, which is a comparison at the scale of 1 to 400 in terms of developers and 1 to millions in comparison to market cap.

I completely agree with you on the choosing where you compete and where you don't. But you still need a product feature comparison in order to do that. The same is true whether you compete in the database market or BIOS market - same rules apply. I work for a CSP in my day job and we have to do feature comparisons with e.g. AWS S3, and obviously we cannot compete, nor should we try to. But as I said - the product feature comparison is your first step of helping you decide what you should have, would maybe like to have, and what you absolutely should not waste resources on.

**Comment by pietrushnic** [comment]
> The same is true whether you compete in the database market or BIOS market - same rules apply. I work for a CSP in my day job and we have to do feature comparisons with e.g. AWS S3, and obviously we cannot compete, nor should we try to.

It boils down to how you do it. We don't have a marketing department, and most engineers don't want to waste their precious time writing marketing documents or comparison tables. I agree it is useful, but only if you can prove something. This is a good discussion for [Dasharo Developers vPub]() and easily getting off-topic here (sorry, my fault).

Please check this slide presented at FOSSASIA 2024:
![dasharo_comparison_table](https://github.com/user-attachments/assets/a7a8a185-ebb5-4305-a5f9-76b54e0d5e33)

**Comment by xmhd** [comment]
Seems there is a patch [1] available upstream for this. When I have some downtime over Christmas/New Year I will do a test build of this and report back.

[1] https://review.coreboot.org/c/coreboot/+/73790

**Comment by macpijan** [comment]
It should be already live in the latest release https://docs.dasharo.com/variants/msi_z690/releases/#v114-2024-11-22

**Comment by xmhd** [comment]
> It should be already live in the latest release https://docs.dasharo.com/variants/msi_z690/releases/#v114-2024-11-22

Perfect, I will test this out and report back. Thanks!

**Comment by xmhd** [comment]
> It should be already live in the latest release https://docs.dasharo.com/variants/msi_z690/releases/#v114-2024-11-22

Any reason why DTS firmware update is not showing me this latest release?

**Comment by xmhd** [comment]
Tested and confirmed working on v0.9.2.

Thanks!
